### PR TITLE
Governance parameters for Dynamic gas price policy

### PR DESCRIPTION
**Pull request by kjeom**
## Proposed changes

- Governance parameters added for dynamic gas price policy
  - LowerBoundBaseFee, UpperBoundBaseFee, GasTarget, BlockGasLimit, BaseFeeDenominator
- Once a related vote submitted, the changed parameter value is written in blockchain chainConfig by callback function after epoch

## Types of changes

Please put an x in the boxes related to your change.

- [ ] Bugfix
- [x] New feature or enhancement
- [ ] Others

## Checklist

*Put an x in the boxes that apply. You can also fill these out after creating the PR. If you're unsure about any of them, don't hesitate to ask. We're here to help! This is simply a reminder of what we are going to look for before merging your code.*

- [x] I have read the [CONTRIBUTING GUIDELINES](https://github.com/klaytn/klaytn/blob/master/CONTRIBUTING.md) doc
- [x] I have signed the [CLA](https://cla-assistant.io/klaytn/klaytn)
- [x] Lint and unit tests pass locally with my changes (`$ make test`)
- [x] I have added tests that prove my fix is effective or that my feature works
- [ ] I have added necessary documentation (if appropriate)
- [ ] Any dependent changes have been merged and published in downstream modules

## Related issues

- #1245

### CODE REVIEW

File: params/config.go [hunk 1]
```diff
@@ -201,6 +201,11 @@ type RewardConfig struct {
 	StakingUpdateInterval  uint64   `json:"stakingUpdateInterval"`  // Interval when staking information is updated
 	ProposerUpdateInterval uint64   `json:"proposerUpdateInterval"` // Interval when proposer information is updated
 	MinimumStake           *big.Int `json:"minimumStake"`           // Minimum amount of peb to join CCO
+	LowerBoundBaseFee      uint64   `json:"lowerboundbasefee"`      // minimum base fee for dynamic gas price
```
Comment (sirano11) [inline params/config.go:204]:
```suggestion
	LowerBoundBaseFee      uint64   `json:"lowerboundbasefee"`      // Minimum base fee for dynamic gas price
```

File: params/config.go [hunk 2]
```diff
@@ -201,6 +201,11 @@ type RewardConfig struct {
...
+	UpperBoundBaseFee      uint64   `json:"upperboundbasefee"`      // maximum base fee for dynamic gas price
```
Comment (sirano11) [inline params/config.go:205]:
```suggestion
	UpperBoundBaseFee      uint64   `json:"upperboundbasefee"`      // Maximum base fee for dynamic gas price
```

File: params/config.go [hunk 3]
```diff
@@ -201,6 +201,11 @@ type RewardConfig struct {
...
+	GasTarget              uint64   `json:"gastarget"`              // guage parameter increasing or decreasing gas price
```
Comment (sirano11) [inline params/config.go:206]:
```suggestion
	GasTarget              uint64   `json:"gastarget"`              // Gauge parameter increasing or decreasing gas price
```

File: params/config.go [hunk 4]
```diff
@@ -201,6 +201,11 @@ type RewardConfig struct {
...
+	BlockGasLimit          uint64   `json:"blockgaslimit"`          // maximum network and process capacity to allow in a block
```
Comment (sirano11) [inline params/config.go:207]:
```suggestion
	BlockGasLimit          uint64   `json:"blockgaslimit"`          // Maximum network and process capacity to allow in a block
```

File: params/config.go [hunk 5]
```diff
@@ -201,6 +201,11 @@ type RewardConfig struct {
...
+	BaseFeeDenominator     uint64   `json:"basefeedenominator"`     // for nomalizing effect of the rapid change like impulse gas used
```
Comment (sirano11) [inline params/config.go:208]:
```suggestion
	BaseFeeDenominator     uint64   `json:"basefeedenominator"`     // For normalizing effect of the rapid change like impulse gas used
```

File: governance/default.go [hunk 6]
```diff
@@ -81,6 +86,11 @@ var (
 		params.Policy:                  "istanbul.policy",
 		params.CommitteeSize:           "istanbul.committeesize",
 		params.UnitPrice:               "governance.unitprice",
+		params.LowerBoundBaseFee:       "reward.lowerboundbasefee",
```
Comment (sirano11) [inline governance/default.go:89]: How about renaming key?

File: governance/handler.go [hunk 7]
```diff
@@ -65,6 +70,26 @@ var GovernanceItems = map[int]check{
...
+}
+
+func updateBlockGasLimit(g *Governance, k string, v interface{}) {
+	g.ChainConfig.Governance.Reward.BlockGasLimit = v.(uint64)
+}
+
+func updateBaseFeeDenominator(g *Governance, k string, v interface{}) {
+	g.ChainConfig.Governance.Reward.BaseFeeDenominator = v.(uint64)
+}
+
```
Comment (blukat29) [inline governance/handler.go:92]: Please do not modify `Governance.ChainConfig`. It is intended to store the initial chain config.

What can change is BlockChain.chainConfig. If you want to edit BlockChain.chainConfig, then you should do similar to https://github.com/klaytn/klaytn/blob/v1.8.3/blockchain/blockchain.go#L429. and https://github.com/klaytn/klaytn/blob/v1.8.3/governance/handler.go#L80.

```go
// blockchain/blockchain.go
func (bc *BlockChain) GasTarget() uint64 { ... }
func (bc *BlockChain) SetGasTarget(val uint64) { ... }

// governance/handler.go
func updateBlockGasLimit(g *Governance, k string, v interface{}) {
	if g.blockChain != nil {
		g.blockChain.SetGasTarget(g.GetGovernanceValue(params.GasTarget).(uint64))
	}
}
```

**Review by blukat29 — CHANGES_REQUESTED** [review]

**Comment by kjeom** [comment]
closed by corrupted rebase

**Pull request closed without merging**